File: Jal/Al/src/C4/U3/PrimMST.java
package C4.U3;

import C2.U4.IndexMinPQ;
import C1.Unit3_Stacks.Queue;

public class PrimMST {
    // shortest edge from tree vertex
    private Edge[] edgeTo;
    // distTo[w] = edgeTo[w].weight()
    private double[] distTo;
    // true if v on tree
    private boolean[] marked;
    // eligible crossing edges
    private IndexMinPQ<Double> pq;

    private Queue<Edge> mst;

    private double weight;

    public PrimMST(EdgeWeightedGraph G)
    {
        edgeTo = new Edge[G.V()];
        distTo = new double[G.V()];
        marked = new boolean[G.V()];
        mst = new Queue<Edge>();
        for(int v = 0; v < G.V(); v++)
            distTo[v] = Double.POSITIVE_INFINITY;
        pq = new IndexMinPQ<Double>(G.V());

        distTo[0] = 0.0;
        // Initialize pq with 0, weight 0.
        pq.insert(0, 0.0);
        visit(G, pq.delMin());
        while(!pq.isEmpty())
        {
            int VertexMin = pq.min();
            mst.enqueue(edgeTo[VertexMin]);
            weight += edgeTo[VertexMin].weight();
            // Add closed vertex to tree.
            visit(G, pq.delMin());

        }
    }

    private void visit(EdgeWeightedGraph G, int v)
    {
        // Add v to tree; update data structures.
        marked[v] = true;
        for(Edge e : G.adj(v))
        {
            int w = e.other(v);
            // v-w is ineligible.
            if(marked[w])
                continue;
            if(e.weight() < distTo[w])
            {
                // Edge e is new best connection from tree to w.
                edgeTo[w] = e;
                distTo[w] = e.weight();
                if(pq.contains(w))
                    pq.change(w, distTo[w]);
                else
                    pq.insert(w, distTo[w]);
            }
        }
    }

    public Iterable<Edge> edges()
    {
        return mst;
    }

    public double weight()
    {
       return weight;
    }

    public static void main(String[] args)
    {
        StdLib.In in = new StdLib.In(args[0]);
        EdgeWeightedGraph G;
        G = new EdgeWeightedGraph(in);

        PrimMST mst = new PrimMST(G);
        for(Edge e : mst.edges())
            StdLib.StdOut.println(e);
        StdLib.StdOut.println(mst.weight());
    }

}

File: Cal/Al/src/C4/U3/PrimMST.cpp
#include "C4/U3/PrimMST.h"
#include "util/CDouble.h"

#include <cmath>

using namespace std;

PrimMST::PrimMST(EdgeWeightedGraph* G)
{
    edgeTo.reserve(G->getV());
    distTo.reserve(G->getV());
    marked.reserve(G->getV());
    for(int v = 0; v < G->getV(); v++)
    {
        edgeTo.push_back(nullptr);
        /* distTo.push_back(std::numeric_limits<double>::infinity()); */

        distTo.push_back(HUGE_VAL);
        marked.push_back(false);
    }
    pq = new IndexMinPQ(G->getV());

    distTo[0] = 0.0;
    // Initial pq with 0, weight 0.
    pq->insert(0, new CDouble(0.0));
    visit(G, pq->delMin());
    while(!pq->isEmpty())
    {
        int VertexMin = pq->min();
        mst.enqueue(edgeTo[VertexMin]);
        weight += edgeTo[VertexMin]->getWeight();
        // Add closed vertex to tree.
        visit(G, pq->delMin());
    }
}

void PrimMST::visit(EdgeWeightedGraph* G, int v)
{
    // Add v to tree; update data structures.
    marked[v] = true;
    vector<Edge*> myEdges;
    G->getAdj(v, myEdges);
    for(vector<Edge*>::iterator it = myEdges.begin(); it != myEdges.end(); it++)
    {
        int w = (*it)->other(v);
        // v-w is ineligible.
        if(marked[w])
            continue;
        if((*it)->getWeight() < distTo[w])
        {
            // Edge e is new best connection form tree to w.
            edgeTo[w] = *it;
            distTo[w] = (*it)->getWeight();
            if(pq->contains(w))
                pq->change(w, new CDouble(distTo[w]));
            else
                pq->insert(w, new CDouble(distTo[w]));
        }
    }
}

void PrimMST::getEdges(Queue<Edge*>& qEdges)
{
    while(!mst.isEmpty())
    {
        qEdges.enqueue(mst.dequeue());
    }
}

double PrimMST::getWeight()
{
    return weight;
}
